datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+                # The latter might be dumb and without the execution of a
+                # post-update hook we might not be able to retrieve the
+                # server-side git-annex branch updates (and git-annex does
+                # not trigger the hook on copy), but we know we have
+                # full access via the push url -- we have just used it to copy.
+                fetch_cmd = ['fetch', target, 'git-annex']
+                pushurl = repo.config.get(
+                    'remote.{}.pushurl'.format(target), None)
+                if pushurl:
+                    # for some reason overwriting remote.{target}.url
+                    # does not have any effect...
+                    fetch_cmd = [
+                        '-c',
+                        'url.{}.insteadof={}'.format(
+                            pushurl,
+                            repo.config.get(
+                                'remote.{}.url'.format(target), None)
+                        )
+                    ] + fetch_cmd
+                    lgr.debug(
+                        "Sync local annex branch from pushurl after remote "
+                        'availability update.')
+                repo.call_git(fetch_cmd)
+                _sync_annex(repo, target)
+            except CommandError as e:
+                # it is OK if the remote doesn't have a git-annex branch yet
+                # (e.g. fresh repo)
+                # TODO is this possible? we just copied? Maybe check if anything
+                # was actually copied?
+                if "fatal: couldn't find remote ref git-annex" not in e.stderr:
+                    raise
+                lgr.debug('Remote does not have a git-annex branch: %s', e)
+            # and push the annex branch to announce local availability info
+            # too
+            yield from _push_refspecs(
+                repo,
+                target,
+                ['git-annex'
+                 if ds.config.get('branch.git-annex.merge', None)
+                 else 'git-annex:git-annex'],
+                force,
+                res_kwargs.copy(),
+            )
+
+
+def _sync_annex(repo, target, corresponding_branch=None):
+    if not corresponding_branch:
+        corresponding_branch = repo.get_corresponding_branch()
+    synced_branch = 'synced/{}'.format(corresponding_branch)
+    had_synced_branch = synced_branch in repo.get_branches()
+    repo.call_git([
+        'annex', 'sync',
+        # there is no way to do this in a purely local fashion
+        # although this is what we want here, emulate...
+        target,
+        # disable any external interaction and other magic
+        '--no-push', '--no-pull', '--no-commit', '--no-resolvemerge',
+    ])
+    if not had_synced_branch:
+        # cleanup after sync, simply remove the sync branch and not
+        # use 'sync --cleanup' to further limit remote interaction
+        repo.call_git(['branch', '-D', synced_branch])
```

Comment:
This seems to be duplicating a good amount of logic from `AnnexRepo._save_post`.  It'd be good to consolidate it, or put in a comment explaining why that's not a good idea.

Reply:
True, will do!

Reply:
Came in with #4243